aten/src/ATen/core/op_registration/adaption.h

```diff
@@ -57,5 +60,45 @@ check_tensor_options_and_extract_memory_format(
     return options.memory_format_opt();
   }
 }
+
+inline void check_or_update_common_device(optional<Device>& common_device, const at::Tensor& tensor, at::CheckedFrom c) {
+  if (!tensor.defined()) {
+    return;
+  }
+
+  if (tensor.is_cpu() && tensor.dim() == 0) {
+    // CPU 0-dimenstion tensor is not considered as common device
+    return;
+  }
+
+  if (!common_device.has_value()) {
+    common_device = tensor.device();
+    return;
+  }
+
+  TORCH_CHECK(
```

Comment:
@bhosmer : Thanks! I also note in that case we probably don't want to make a call each time to do the check (since in most case the happy path shouldn't have common device check failure).

So... I am tempted to put the check method just throw, e.g. 
```
void common_device_check_failure() {
   TORCH_CHECK(false, "Expected all tensors...")
}
```

Hope this is not an anti-pattern....

Reply:
Ah good point - in fact it might be worth wrapping the check expression in a `C10_UNLIKELY` as well. I don't think it's an antipattern - but cc @swolchok who's recently done some similar refactoring to get throws out of inline-worthy functions to for a best-practice check 😁